### ISSUE TITLE
[New device support]: Radiator Valve TS0601 - Created a converter but im missing some exposed topics and other are not quite working.

### ISSUE THREAD

**Issue by dweng0** [Koenkk/zigbee2mqtt#23772]
### Link

https://www.aliexpress.com/item/1005006139544669.html?spm=a2g0o.order_list.order_list_main.53.1acb1802tbDjWe

### Database entry
``` javascript
{   "id": 17,   "type": "EndDevice",   "ieeeAddr": "0xa4c138a93ec7382b",   "nwkAddr": 30083,   "manufId": 4417,   "manufName": "_TZE200_yqgbrdyo",   "powerSource": "Battery",   "modelId": "TS0601",   "epList": [     1   ],   "endpoints": {     "1": {       "profId": 260,       "epId": 1,       "devId": 81,       "inClusterList": [         4,         5,         61184,         0       ],       "outClusterList": [         25,         10       ],       "clusters": {         "genBasic": {           "attributes": {             "modelId": "TS0601",             "manufacturerName": "_TZE200_yqgbrdyo",             "powerSource": 3,             "zclVersion": 3,             "appVersion": 67,             "stackVersion": 0,             "hwVersion": 1,             "dateCode": ""           }         }       },       "binds": [],       "configuredReportings": [],       "meta": {}     }   },   "appVersion": 67,   "stackVersion": 0,   "hwVersion": 1,   "dateCode": "",   "zclVersion": 3,   "interviewCompleted": true,   "meta": {},   "lastSeen": 1724770412186 }{   "id": 17,   "type": "EndDevice",   "ieeeAddr": "0xa4c138a93ec7382b",   "nwkAddr": 30083,   "manufId": 4417,   "manufName": "_TZE200_yqgbrdyo",   "powerSource": "Battery",   "modelId": "TS0601",   "epList": [     1   ],   "endpoints": {     "1": {       "profId": 260,       "epId": 1,       "devId": 81,       "inClusterList": [         4,         5,         61184,         0       ],       "outClusterList": [         25,         10       ],       "clusters": {         "genBasic": {           "attributes": {             "modelId": "TS0601",             "manufacturerName": "_TZE200_yqgbrdyo",             "powerSource": 3,             "zclVersion": 3,             "appVersion": 67,             "stackVersion": 0,             "hwVersion": 1,             "dateCode": ""           }         }       },       "binds": [],       "configuredReportings": [],       "meta": {}     }   },   "appVersion": 67,   "stackVersion": 0,   "hwVersion": 1,   "dateCode": "",   "zclVersion": 3,   "interviewCompleted": true,   "meta": {},   "lastSeen": 1724770412186 }
```
## Comments
So i have managed to get it somewhat working buy using existing definitions and pulling in bits that work and cross referencing it with the manual.

Using system mode, if i set (between heat, off and auto) i get the following error
``` bash
z2m: EventBus error 'Publish/mqttMessage': Cannot read properties of undefined (reading 'some')
z2m: Publish 'set' 'current_heating_setpoint' to '0xa4c138a93ec7382b' failed: 'Error: ZCL command 0xa4c138a93ec7382b/1 manuSpecificTuya.dataRequest({"seq":1,"dpValues":[{"dp":4,"datatype":2,"data":[0,0,0,220]}]}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"reservedBits":0,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240))'
z2m: EventBus error 'Publish/mqttMessage': Cannot read properties of undefined (reading 'some')
z2m: Publish 'set' 'system_mode' to '0xa4c138a93ec7382b' failed: 'Error: No datapoint defined for 'system_mode''
```
But the read only version `running_state` is showing the correct state (idle if not heating and, heating if heating..). I took the datapoint for this [here](https://github.com/Koenkk/zigbee-herdsman-converters/blob/8e3bb084a7d005affccb0d7ba28fa70be28bb494/src/devices/tuya.ts#L4621)

### External definition

```shell
const definition = {
    fingerprint: [
        {
            modelID: 'TS0601',
            manufacturerName: '_TZE200_yqgbrdyo',
        },
    ],
    model: 'TRV06',
    vendor: 'EARU-electric',
    description: 'TRV',
   fromZigbee: [tuya.fz.datapoints],
        toZigbee: [tuya.tz.datapoints],     
        onEvent: tuya.onEventSetTime,
        configure: tuya.configureMagicPacket,
        exposes: [
            e.child_lock(),
            e.battery_low(),
            e.window_detection(),
            e
                .climate()
                .withSetpoint('current_heating_setpoint', 5, 35, 1, ea.STATE_SET)
                .withLocalTemperature(ea.STATE)
                .withSystemMode(['auto', 'heat', 'off'], ea.STATE_SET)
                .withRunningState(['idle', 'heat'], ea.STATE)
                .withLocalTemperatureCalibration(-9, 9, 1, ea.STATE_SET),
            ...tuya.exposes.scheduleAllDays(ea.STATE_SET, 'HH:MM/C HH:MM/C HH:MM/C HH:MM/C HH:MM/C HH:MM/C'),
            e
                .binary('scale_protection', ea.STATE_SET, 'ON', 'OFF')
                .withDescription(
                    'If the heat sink is not fully opened within ' +
                        'two weeks or is not used for a long time, the valve will be blocked due to silting up and the heat sink will not be ' +
                        'able to be used. To ensure normal use of the heat sink, the controller will automatically open the valve fully every ' +
                        'two weeks. It will run for 30 seconds per time with the screen displaying "Ad", then return to its normal working state ' +
                        'again.',
                ),
            e
                .binary('frost_protection', ea.STATE_SET, 'ON', 'OFF')
                .withDescription(
                    'When the room temperature is lower than 5 °C, the valve opens; when the temperature rises to 8 °C, the valve closes.',
                ),
            e.numeric('error', ea.STATE).withDescription('If NTC is damaged, "Er" will be on the TRV display.'),
        ],
        meta: {
            tuyaDatapoints: [
                [35, 'battery_low', tuya.valueConverter.trueFalse0],
                [3, 'running_state', tuya.valueConverterBasic.lookup({heat: tuya.enum(0), idle: tuya.enum(1)})],
                [4, 'current_heating_setpoint', tuya.valueConverter.divideBy10],
                [5, 'local_temperature', tuya.valueConverter.divideBy10],
                [7, 'child_lock', tuya.valueConverter.lockUnlock],
                [14, 'window_detection', tuya.valueConverter.onOff],
                [8, 'open_window', tuya.valueConverter.onOff],
                [17, 'open_window_time', tuya.valueConverter.raw],
                [36, 'frost_protection', tuya.valueConverter.onOff],
                [39, 'scale_protection', tuya.valueConverter.onOff],
                [47, 'local_temperature_calibration', tuya.valueConverter.localTempCalibration2],
            ],
        },
};

module.exports = definition;
```

### What does/doesn't work with the external definition?

I primarily used [this](https://github.dev/Koenkk/zigbee-herdsman-converters/blob/8e3bb084a7d005affccb0d7ba28fa70be28bb494/src/devices/tuya.ts#L4399-L4400) definition for reference, but where any parts were missing, i copied from other reasonable locations.

## What works
The read state for local temperature and running state are working:
![image](https://github.com/user-attachments/assets/3b3f735d-5891-4b2d-91ff-6fc9234a6b1e)
![image](https://github.com/user-attachments/assets/e6b39e22-60b2-436a-9276-124bc82803b2)
The **read** state for child_lock appears to work, but setting it gives an error.

### local_temperature
The **read** state for  **Current heating setpoint** works. If i change the temp on the device, it updates the slider.  However if i set it in the front end, I get the following error:
``` bash
z2m: Publish 'set' 'current_heating_setpoint' to '0xa4c138a93ec7382b' failed: 'Error: ZCL command 0xa4c138a93ec7382b/1 manuSpecificTuya.dataRequest({"seq":1,"dpValues":[{"dp":4,"datatype":2,"data":[0,0,0,180]}]}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"reservedBits":0,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240))'
```
### System_mode
Using system mode, if i set (between heat, off and auto) i get the following error
``` bash
z2m: EventBus error 'Publish/mqttMessage': Cannot read properties of undefined (reading 'some')
z2m: Publish 'set' 'current_heating_setpoint' to '0xa4c138a93ec7382b' failed: 'Error: ZCL command 0xa4c138a93ec7382b/1 manuSpecificTuya.dataRequest({"seq":1,"dpValues":[{"dp":4,"datatype":2,"data":[0,0,0,220]}]}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"reservedBits":0,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240))'
z2m: EventBus error 'Publish/mqttMessage': Cannot read properties of undefined (reading 'some')
z2m: Publish 'set' 'system_mode' to '0xa4c138a93ec7382b' failed: 'Error: No datapoint defined for 'system_mode''
```
But the read only version `running_state` is showing the correct state (idle if not heating and, heating if heating..). It also updates correctly. I took the datapoint for this [here](https://github.com/Koenkk/zigbee-herdsman-converters/blob/8e3bb084a7d005affccb0d7ba28fa70be28bb494/src/devices/tuya.ts#L4621)

### child lock
``` bash
0z2m: Publish 'set' 'child_lock' to '0xa4c138a93ec7382b' failed: 'Error: ZCL command 0xa4c138a93ec7382b/1 manuSpecificTuya.dataRequest({"seq":1,"dpValues":[{"dp":7,"datatype":1,"data":[1]}]}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"reservedBits":0,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240))'
error 2024-08-29 09:37:20z2m: EventBus error 'Publish/mqttMessage': Cannot read properties of undefined (reading 'some')
```
### Scale protection
Gives an error when being set:
```bash
 Publish 'set' 'scale_protection' to '0xa4c138a93ec7382b' failed: 'Error: ZCL command 0xa4c138a93ec7382b/1 manuSpecificTuya.dataRequest({"seq":1,"dpValues":[{"dp":39,"datatype":1,"data":[1]}]}, 
{"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"reservedBits":0,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240))'
```

### Frost protection 
Gives an error when being set:
```bash
z2m: Publish 'set' 'frost_protection' to '0xa4c138a93ec7382b' failed: 'Error: ZCL command 0xa4c138a93ec7382b/1 manuSpecificTuya.dataRequest({"seq":1,"dpValues":[{"dp":36,"datatype":1,"data":[0]}]}, 

{"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"reservedBits":0,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240))'
```

### Temp callibration
Appears to read correctly, but gives an error when being set
``` bash
z2m: Publish 'set' 'local_temperature_calibration' to '0xa4c138a93ec7382b' failed: 'Error: ZCL command 0xa4c138a93ec7382b/1 manuSpecificTuya.dataRequest({"seq":1,"dpValues":[{"dp":47,"datatype":2,"data":[255,255,255,255]}]}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"reservedBits":0,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240))'
```
According to the product manual, there is a programmable mode, but the different "days of the week" i've tried, dont read or write correctly. 

### Low Battery
This just doesn't work, not sure what the read property would be for this.

When i select "**state**" in the tab menu, i get the following:
![image](https://github.com/user-attachments/assets/f5049258-e33e-4148-b07c-4cc600708b46)

**Comment by krecik** [comment]
Please check: https://github.com/Koenkk/zigbee2mqtt/issues/23744

This is probably the same device as: `_TZE200_p3dbf6qs`

**Comment by xb441** [comment]
Hello,
I have the same problem. 
My thermostat valve from Tuya Zigbee Manufacturer **_TZE200_hue3yfsn** work fine.

But my new thermostat valves from Tuya Zigbee Manufacturer **_TZE200_yqgbrdyo** are unsupported

Thank you to the developpers to add this new model to z2m. And many thx for your excellent job.
Xavier

Link https://fr.aliexpress.com/item/1005006139544669.html

generated_external_definition :

__const definition = {
    zigbeeModel: ['TS0601'],
    model: 'TS0601',
    vendor: '_TZE200_yqgbrdyo',
    description: 'Automatically generated definition',_
    extend: [],
    meta: {},
};
module.exports = definition;_

**Comment by xb441** [comment]
> this should be updated on the edge version of z2m

Thx, but I don't use the edge version, but the stable version embedded in the ZigbeeLinger plugin in Jeedom.
So, I have to wait the new version, I hope before winter!!
Xavier

**Comment by xb441** [comment]
many many thanks.
I tried your method and it seems to work well.
I changed the model and the seller in the configuration file 
I put model: 'TS0601_thermostat_3', vendor: 'Tuya', to match what I purchased (the visual of thermostat_3 is exactly the same). This model is already known by Zigbee2Mqtt.

https://www.zigbee2mqtt.io/devices/TS0601_thermostat_3.html#tuya-ts0601_thermostat_3

Should I leave the "external_converters" pointer in the configuration.yaml file as well as the file itself in the directory ?
It seems so?
